File: rust/src/read_file.rs
use std::fs::File;
use std::io::{BufRead, BufReader, Lines};

use flate2::read::GzDecoder;

pub fn get_lines_in_file(pathname: &str) -> Lines<Box<dyn BufRead>> {
    let reader = get_file_reader(pathname);
    reader.lines()
}

// https://stackoverflow.com/questions/36088116/how-to-do-polymorphic-io-from-either-a-file-or-stdin-in-rust
fn get_file_reader(pathname: &str) -> Box<dyn BufRead> {
    println!("Init file: {}", pathname);
    let file = File::open(pathname).unwrap();
    let result: Box<dyn BufRead> = match pathname.ends_with(".gz") {
        true => Box::new(BufReader::new(GzDecoder::new(file))),
        false => Box::new(BufReader::new(file)),
    };
    result
}

File: python/src/read_file.py
from typing import Iterator
import gzip


class FileReader:
    def get_lines_in_pathname(self, pathname: str) -> Iterator[str]:
        print(f"Init file: {pathname}")
        for line in self._get_lines_in_pathname(pathname):
            yield line

    def _get_lines_in_pathname(self, pathname: str) -> Iterator[str]:
        return (
            self._get_lines_in_gz_file(pathname)
            if pathname.endswith(".gz")
            else self._get_lines_in_log_file(pathname)
        )

    def _get_lines_in_log_file(self, pathname: str) -> Iterator[str]:
        with open(pathname, "r") as file:
            for line in file.read().splitlines():
                yield line

    def _get_lines_in_gz_file(self, pathname: str) -> Iterator[str]:
        with gzip.open(pathname, mode="rt") as fp:
            for line in fp:
                yield line
